wolff/optimization/constrained_gradient_descent_optimizer.py

```diff
@@ -0,0 +1,166 @@
+'''Gradient Descent Step'''
+from wolff.solver import Solver
+import numpy as np
+from numpy import linalg as LA
+
+
+class ConstrainedGradientDescentOptimizer(Solver):
+
+    def __init__(self,
+                 dtheta=0.1,
+                 theta_0=0.0,
+                 grids_true=[],
+                 elapsed_times_true=[],
+                 theta_min=-np.inf,
+                 theta_max=np.inf,
+                 epsilon=1e-6,
+                 max_iters=1000,
+                 learning_rate=0.01,
+                 momentum_rate=0.9):
+        super(ConstrainedGradientDescentOptimizer, self).__init__()
+
+        self.dtheta = dtheta
+        self.theta_0 = theta_0
+        self.grids_true = grids_true
+        self.elapsed_times_true = elapsed_times_true
+        self.theta_min = theta_min
+        self.theta_max = theta_max
+        self.epsilon = epsilon
+        self.max_iters = max_iters
+        self.learning_rate = learning_rate
+        self.momentum_rate = momentum_rate
+
+    @property
+    def dtheta(self):
+        return self._dtheta
+
+    @dtheta.setter
+    def dtheta(self, value):
+        self._dtheta = value
+
+    @property
+    def theta_0(self):
+        return self._theta_0
+
+    @theta_0.setter
+    def theta_0(self, value):
+        self._theta_0 = value
+
+    @property
+    def grids_true(self):
+        return self._grids_true
+
+    @grids_true.setter
+    def grids_true(self, value):
+        self._grids_true = value
+
+    @property
+    def elapsed_times_true(self):
+        return self._elapsed_times_true
+
+    @elapsed_times_true.setter
+    def elapsed_times_true(self, value):
+        self._elapsed_times_true = value
```

Comment:
Should we deep copy the list?
```suggestion
        self._elapsed_times_true = list(value)
```